src/Common/MemoryTracker.cpp

```diff
@@ -172,65 +172,81 @@ void MemoryTracker::alloc(Int64 size)
-        DB::TraceCollector::collect(DB::TraceType::MemorySample, StackTrace(), size);
-    }
+        if (unlikely(current_profiler_limit && will_be > current_profiler_limit))
+        {
+            BlockerInThread untrack_lock(VariableContext::Global);
+            DB::TraceCollector::collect(DB::TraceType::Memory, StackTrace(), size);
+            setOrRaiseProfilerLimit((will_be + profiler_step - 1) / profiler_step * profiler_step);
+        }
 
-    if (unlikely(current_hard_limit && will_be > current_hard_limit) && memoryTrackerCanThrow(level, false))
-    {
-        /// Prevent recursion. Exception::ctor -> std::string -> new[] -> MemoryTracker::alloc
-        BlockerInThread untrack_lock(VariableContext::Global);
+        std::bernoulli_distribution sample(sample_probability);
+        if (unlikely(sample_probability && sample(thread_local_rng)))
+        {
+            BlockerInThread untrack_lock(VariableContext::Global);
+            DB::TraceCollector::collect(DB::TraceType::MemorySample, StackTrace(), size);
+        }
 
-        ProfileEvents::increment(ProfileEvents::QueryMemoryLimitExceeded);
-        const auto * description = description_ptr.load(std::memory_order_relaxed);
-        amount.fetch_sub(size, std::memory_order_relaxed);
-        throw DB::Exception(DB::ErrorCodes::MEMORY_LIMIT_EXCEEDED,
-                            "Memory limit{}{} exceeded: would use {} (attempt to allocate chunk of {} bytes), maximum: {}",
-                            description ? " " : "", description ? description : "",
-                            formatReadableSizeWithBinarySuffix(will_be),
-                            size, formatReadableSizeWithBinarySuffix(current_hard_limit));
+        if (unlikely(current_hard_limit && will_be > current_hard_limit) && memoryTrackerCanThrow(level, false))
+        {
+            /// Prevent recursion. Exception::ctor -> std::string -> new[] -> MemoryTracker::alloc
+            BlockerInThread untrack_lock(VariableContext::Global);
+
+            /// TODO: Should we increment limit exceeded event if throw_if_memory_exceeded = true.
+            /// It will be incremented during next call where it will be safe to throw exception
+            ProfileEvents::increment(ProfileEvents::QueryMemoryLimitExceeded);
+            const auto * description = description_ptr.load(std::memory_order_relaxed);
+            amount.fetch_sub(size, std::memory_order_relaxed);
+            throw DB::Exception(DB::ErrorCodes::MEMORY_LIMIT_EXCEEDED,
+                                "Memory limit{}{} exceeded: would use {} (attempt to allocate chunk of {} bytes), maximum: {}",
+                                description ? " " : "", description ? description : "",
+                                formatReadableSizeWithBinarySuffix(will_be),
+                                size, formatReadableSizeWithBinarySuffix(current_hard_limit));
+        }
     }
 
     updatePeak(will_be);
```

Comment:
There is logging in `updatePeak` that should be disabled in "light" version of MemoryTracker.

Reply:
Fixed.